Sources/NIO/BSDSocket.swift

```diff
@@ -0,0 +1,897 @@
+      throws {
+#if os(Windows)
+    if WinSDK.getsockname(s, name, namelen) == SOCKET_ERROR {
+      throw IOError(winsock: WSAGetLastError(), reason: "getsockname")
+    }
+#else
+#if os(iOS) || os(macOS) || os(tvOS) || os(watchOS)
+    let getsockname: (CInt, UnsafeMutablePointer<sockaddr>, UnsafeMutablePointer<socklen_t>) -> CInt = Darwin.getsockname
+#else
+    let getsockname: (CInt, UnsafeMutablePointer<sockaddr>?, UnsafeMutablePointer<socklen_t>?) -> CInt = Glibc.getsockname
+#endif
+    _ = try syscall(blocking: false) {
+      getsockname(s, name, namelen)
+    }
+#endif
+  }
+
+  @inline(never)
+  static func getsockopt(socket: BSDSocket.Handle, level: BSDSocket.OptionLevel,
+                         option_name optname: BSDSocket.Option,
+                         option_value optval: UnsafeMutableRawPointer,
+                         option_len optlen: UnsafeMutablePointer<socklen_t>)
+      throws {
+#if os(Windows)
+    if WinSDK.getsockopt(socket, level.rawValue, optname.rawValue,
+                         optval.assumingMemoryBound(to: CChar.self),
+                         optlen) == SOCKET_ERROR {
+      throw IOError(winsock: WSAGetLastError(), reason: "getsockopt")
+    }
+#else
+#if os(iOS) || os(macOS) || os(tvOS) || os(watchOS)
+    let getsockopt: (CInt, CInt, CInt, UnsafeMutableRawPointer, UnsafeMutablePointer<socklen_t>) -> CInt = Darwin.getsockopt
+#else
+    let getsockopt: (CInt, CInt, CInt, UnsafeMutableRawPointer?, UnsafeMutablePointer<socklen_t>?) -> CInt = Glibc.getsockopt
+#endif
+    _ = try syscall(blocking: false) {
+      getsockopt(socket, level.rawValue, optname.rawValue, optval, optlen)
+    }
+#endif
+  }
+
+  @inline(never)
+  static func listen(socket s: BSDSocket.Handle, backlog: CInt) throws {
+#if os(Windows)
+    if WinSDK.listen(s, backlog) == SOCKET_ERROR {
+      throw IOError(winsock: WSAGetLastError(), reason: "listen")
+    }
+#else
+#if os(iOS) || os(macOS) || os(tvOS) || os(watchOS)
+    let listen: (CInt, CInt) -> CInt = Darwin.listen
+#else
+    let listen: (CInt, CInt) -> CInt = Glibc.listen
+#endif
+    _ = try syscall(blocking: false) {
+      listen(s, backlog)
+    }
+#endif
+  }
+
+  @inline(never)
+  static func recv(socket s: BSDSocket.Handle,
+                   buffer buf: UnsafeMutableRawPointer, length len: size_t)
+      throws -> IOResult<size_t> {
+#if os(Windows)
+    let iResult: CInt =
+        WinSDK.recv(s, buf.assumingMemoryBound(to: CChar.self), CInt(len), 0)
+    if iResult == SOCKET_ERROR {
+      throw IOError(winsock: WSAGetLastError(), reason: "recv")
+    }
+    return .processed(size_t(iResult))
+#else
+#if os(iOS) || os(macOS) || os(tvOS) || os(watchOS)
+    let recv: (CInt, UnsafeMutableRawPointer, size_t, CInt) -> ssize_t = Darwin.recv
+#else
+    let recv: (CInt, UnsafeMutableRawPointer?, size_t, CInt) -> ssize_t = Glibc.recv
+#endif
+    return try syscall(blocking: true) {
+      read(s, buf, len, 0)
```

Comment:
Looks like a typo

Reply:
No, this is not a typo - the `read` call was used previously - but the new wrapping here uses the original BSD API which called it `recv` since it was a sockets-only API rather than a file descriptor based API.

Reply:
Yeah, Im planning on changing the way that this is done, and that should help with that.